Added missing functions

rheslip opened the pull request:
Added in the rest of the Teensy audio functions that are not specific to the Teensy hardware. Also changed default sample rate to 44.1k and default sample block size to 128 for maximum compatibility with Teensy sketches. 96k chews over 2x the CPU vs 44.1k.

Two hooks I wrote a couple years ago are in there as well: effect_daisysp.h and synth_daisySP.h. These facilitate using the DaisySP library with Teensy Audio. See my github for how these work and examples.

rheslip committed:
Completing work started by ghostintranslation

Many changes and additions to bring the initial port up to 95% completion.  Changes sample rate to 44100 from 96k which reduces CPU load by more than 50%. Changed sample block size from 256 to 128 which is the default for the Teensy Audio library and is required for some functions e.g. FFT analysis.

rheslip committed:
Added examples

Added a couple more examples and updated readme with known issues

rheslip committed:
Update README.md

Updated readme - DaisySP functions.

ghostintranslation commented:
Hey thanks for your contribution, sorry I haven't yet reviewed I am busy with other things, but I haven't forgotten. I will review as soon as I can.

rheslip commented:
No problem.

________________________________
From: Ghost In Translation ***@***.***>
Sent: July 2, 2025 2:31 PM
To: ghostintranslation/pico-audio ***@***.***>
Cc: rheslip ***@***.***>; Author ***@***.***>
Subject: Re: [ghostintranslation/pico-audio] Added missing functions (PR #6)

[https://avatars.githubusercontent.com/u/50341468?s=20&v=4]ghostintranslation left a comment (ghostintranslation/pico-audio#6)<https://github.com/ghostintranslation/pico-audio/pull/6#issuecomment-3028938570>

Hey thanks for your contribution, sorry I haven't yet reviewed I am busy with other things, but I haven't forgotten. I will review as soon as I can.

—
Reply to this email directly, view it on GitHub<https://github.com/ghostintranslation/pico-audio/pull/6#issuecomment-3028938570>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABRTUUHKQMRVZVKNQZBXVCL3GQQRLAVCNFSM6AAAAAB56YM6PCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMRYHEZTQNJXGA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

okyeron commented:
In testing these changes I am unable to compile with just the contents of the repo 
(if I do not have the Teensy Audio Library installed as an Arduino Library)

```
In file included from /Users/nevetsokyeron/Documents/GitHub/pico-audio/pico-audio.h:6,
                 from /Users/nevetsokyeron/Documents/GitHub/pico-audio/examples/PolySynth/PolySynth.ino:13:
/Users/nevetsokyeron/Documents/GitHub/pico-audio/analyze_fft256.h:32:10: fatal error: arm_math.h: No such file or directory
   32 | #include <arm_math.h>    // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h
```
   
The various references to this would need to be changed to 
   `#include "arm_math/src/arm_math.h"` to use the included arm_math/h file and not the teensy library version

This will also error if you have the Adafruit port of the Teensy Audio Library installed. `Audio_-_Adafruit_Fork`

Changing references to AudioStream.h to the local copy in this repo would be more portable `#include "AudioStream.h"`

---

After all that I get compile errors around `__enable_irq()` and `__disable_irq` in the example sketches

```
In file included from /Users/nevetsokyeron/Documents/GitHub/pico-audio/pico-audio.h:45,
                 from /Users/nevetsokyeron/Documents/GitHub/pico-audio/examples/SimpleDrum/SimpleDrum.ino:3:
/Users/nevetsokyeron/Documents/GitHub/pico-audio/synth_dc.h: In member function 'void AudioSynthWaveformDc::amplitude(float)':
/Users/nevetsokyeron/Documents/GitHub/pico-audio/synth_dc.h:97:17: error: '__disable_irq' was not declared in this scope
   97 |                 __disable_irq();
      |                 ^~~~~~~~~~~~~
```

it seems that these defines are commented out in the local copy of AudioStream.h
```
//#define __disable_irq(...) noInterrupts()
//#define __enable_irq(...)    interrupts()
```

ghostintranslation commented on README.md:
```diff
@@ -1,9 +1,30 @@
+
+The library currently supports 16 bit I2S audio output only and does not yet support audio input.
+
+This library requires the Arduino CMSIS-DSP library (currently V 5.7.0) to be installed.
```
The lib is in the repo actually, so we shouldn't have to install anything else I believe:
https://github.com/ghostintranslation/pico-audio/tree/main/arm_math

ghostintranslation commented on analyze_fft1024.cpp:
```diff
@@ -0,0 +1,134 @@
+		break;
+	}
+//#else
+	release(block);
```
In the original file, because of the condition it's would be doing either the above code or this `release` , now it's doing both. Should this line be commented too or else replace the condition?

rheslip commented:
The release(block); should be commented out too. I commented out the architecture test because Pico 2 is M3, not M4. Ideally the conditional should test for M3 or M4 architecture.

________________________________
From: Ghost In Translation ***@***.***>
Sent: July 14, 2025 6:25 PM
To: ghostintranslation/pico-audio ***@***.***>
Cc: rheslip ***@***.***>; Author ***@***.***>
Subject: Re: [ghostintranslation/pico-audio] Added missing functions (PR #6)


@ghostintranslation commented on this pull request.

________________________________

In analyze_fft1024.cpp<https://github.com/ghostintranslation/pico-audio/pull/6#discussion_r2205927109>:

> +                     output[i] = sqrt_uint32_approx(magsq);
+               }
+               outputflag = true;
+               release(blocklist[0]);
+               release(blocklist[1]);
+               release(blocklist[2]);
+               release(blocklist[3]);
+               blocklist[0] = blocklist[4];
+               blocklist[1] = blocklist[5];
+               blocklist[2] = blocklist[6];
+               blocklist[3] = blocklist[7];
+               state = 4;
+               break;
+       }
+//#else
+       release(block);


In the original file, because of the condition it's would be doing either the above code or this release , now it's doing both. Should this line be commented too or else replace the condition?

—
Reply to this email directly, view it on GitHub<https://github.com/ghostintranslation/pico-audio/pull/6#pullrequestreview-3017984568>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABRTUUH2CAZE7R6BHEPF6MT3IQU4DAVCNFSM6AAAAAB56YM6PCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMJXHE4DINJWHA>.
You are receiving this because you authored the thread.

[https://s-install.avcdn.net/ipm/preview/icons/icon-envelope-tick-green-avg-v1.png]<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>     Virus-free.www.avg.com<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>

rheslip commented:
"it seems that these defines are commented out in the local copy of AudioStream.h

//#define __disable_irq(...) noInterrupts()
//#define __enable_irq(...)    interrupts()

"

As I recall it would not compile with these macros
________________________________
From: Steven Noreyko ***@***.***>
Sent: July 14, 2025 7:46 AM
To: ghostintranslation/pico-audio ***@***.***>
Cc: rheslip ***@***.***>; Author ***@***.***>
Subject: Re: [ghostintranslation/pico-audio] Added missing functions (PR #6)

[https://avatars.githubusercontent.com/u/1909661?s=20&v=4]okyeron left a comment (ghostintranslation/pico-audio#6)<https://github.com/ghostintranslation/pico-audio/pull/6#issuecomment-3069182639>

In testing these changes I am unable to compile with just the contents of the repo
(if I do not have the Teensy Audio Library installed as an Arduino Library)

In file included from /Users/nevetsokyeron/Documents/GitHub/pico-audio/pico-audio.h:6,
                 from /Users/nevetsokyeron/Documents/GitHub/pico-audio/examples/PolySynth/PolySynth.ino:13:
/Users/nevetsokyeron/Documents/GitHub/pico-audio/analyze_fft256.h:32:10: fatal error: arm_math.h: No such file or directory
   32 | #include <arm_math.h>    // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h```

The various references to this would need to be changed to
   `#include "arm_math/src/arm_math.h"` to use the included arm_math/h file and not the teensy library version

This will also error if you have the Adafruit port of the Teensy Audio Library installed. `Audio_-_Adafruit_Fork`

Changing references to AudioStream.h to the local copy in this repo would be more portable `#include "AudioStream.h"`

---

After all that I get compile errors around `__enable_irq()` and `__disable_irq` in the example sketches



In file included from /Users/nevetsokyeron/Documents/GitHub/pico-audio/pico-audio.h:45,
from /Users/nevetsokyeron/Documents/GitHub/pico-audio/examples/SimpleDrum/SimpleDrum.ino:3:
/Users/nevetsokyeron/Documents/GitHub/pico-audio/synth_dc.h: In member function 'void AudioSynthWaveformDc::amplitude(float)':
/Users/nevetsokyeron/Documents/GitHub/pico-audio/synth_dc.h:97:17: error: '__disable_irq' was not declared in this scope
97 | __disable_irq();
| ^~~~~~~~~~~~~```

it seems that these defines are commented out in the local copy of AudioStream.h

//#define __disable_irq(...) noInterrupts()
//#define __enable_irq(...)    interrupts()


—
Reply to this email directly, view it on GitHub<https://github.com/ghostintranslation/pico-audio/pull/6#issuecomment-3069182639>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABRTUUALQMXAS72DMZ7WLZD3IOKCXAVCNFSM6AAAAAB56YM6PCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTANRZGE4DENRTHE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

rheslip commented:
I did try to use the code that's in the repo but could not get it to link. Using the DSP library got it working so I went with that. I am not much of a programmer and not great at debugging toolchain issues either so I end up brute forcing things to make them work.
________________________________
From: Ghost In Translation ***@***.***>
Sent: July 14, 2025 5:36 PM
To: ghostintranslation/pico-audio ***@***.***>
Cc: rheslip ***@***.***>; Author ***@***.***>
Subject: Re: [ghostintranslation/pico-audio] Added missing functions (PR #6)


@ghostintranslation commented on this pull request.

________________________________

In README.md<https://github.com/ghostintranslation/pico-audio/pull/6#discussion_r2205870281>:

>
-Tested only with Pico 2 W so far.
+Note that Teensy 4 and 4.1 are MUCH faster that the RP2350 (approx 5x) so if you need high performance DSP use Teensy 4.
+
+This fork has been modifed from the original 96k/256 sample port to use 44.1k samples per second and 128 sample blocks to maximize compatibilty with Teensy code. Lower samplerate=lower CPU usage.
+
+The library currently supports 16 bit I2S audio output only and does not yet support audio input.
+
+This library requires the Arduino CMSIS-DSP library (currently V 5.7.0) to be installed.


The lib is in the repo actually, so we shouldn't have to install anything else I believe:
https://github.com/ghostintranslation/pico-audio/tree/main/arm_math

—
Reply to this email directly, view it on GitHub<https://github.com/ghostintranslation/pico-audio/pull/6#pullrequestreview-3017903081>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABRTUUGFYSC43HCN3T7FLMD3IQPELAVCNFSM6AAAAAB56YM6PCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMJXHEYDGMBYGE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

ghostintranslation commented on AudioStream.h:
```diff
@@ -41,8 +41,8 @@
 #define ARM_DWT_CYCCNT (m33_hw->dwt_cyccnt)
-#define __disable_irq(...) noInterrupts()
-#define __enable_irq(...)    interrupts()
+//#define __disable_irq(...) noInterrupts()
```
Why removing the macros? With these the code would require less changes to work for the RP2350.

rheslip commented:
As I recall it would not compile.
________________________________
From: Ghost In Translation ***@***.***>
Sent: July 20, 2025 4:35 PM
To: ghostintranslation/pico-audio ***@***.***>
Cc: rheslip ***@***.***>; Author ***@***.***>
Subject: Re: [ghostintranslation/pico-audio] Added missing functions (PR #6)


@ghostintranslation commented on this pull request.

________________________________

In AudioStream.h<https://github.com/ghostintranslation/pico-audio/pull/6#discussion_r2217955053>:

> @@ -41,8 +41,8 @@
 #define __IMXRT1062__  // pretend this is a Teensy 4.x
 #define FLASHMEM
 #define ARM_DWT_CYCCNT (m33_hw->dwt_cyccnt)
-#define __disable_irq(...) noInterrupts()
-#define __enable_irq(...)    interrupts()
+//#define __disable_irq(...) noInterrupts()


Why removing the macros? With these the code would require less changes to work for the RP2350.

—
Reply to this email directly, view it on GitHub<https://github.com/ghostintranslation/pico-audio/pull/6#pullrequestreview-3036172321>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABRTUUG3QKKHEYC5IVQE2TL3JP4RVAVCNFSM6AAAAAB56YM6PCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMZWGE3TEMZSGE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

ghostintranslation commented on effect_fade.cpp:
```diff
@@ -0,0 +1,100 @@
+/* Audio Library for Teensy 3.X
```
This file was changed 10months ago, would be good to have the latest version: https://github.com/PaulStoffregen/Audio/blob/master/effect_fade.cpp

ghostintranslation commented on effect_freeverb.cpp:
```diff
@@ -0,0 +1,493 @@
+		outputR = sat16(bufout - outputR, 1);
+		if (++allpass4indexR >= sizeof(allpass4bufR)/sizeof(int16_t)) allpass4indexR = 0;
+
+		outblockR->data[i] = sat16(outputL * 30, 0);
```
This file was also updated 10 months ago, fixing this mistake.

ghostintranslation commented on effect_granular.h:
```diff
@@ -0,0 +1,65 @@
+/*
```
Similarly this file has an update.

ghostintranslation commented on effect_reverb.cpp:
```diff
@@ -0,0 +1,207 @@
+/*
```
This file has also some updates.

ghostintranslation commented on filter_fir.h:
```diff
@@ -0,0 +1,69 @@
+
+#include <Arduino.h>     // github.com/PaulStoffregen/cores/blob/master/teensy4/Arduino.h
+#include <AudioStream.h> // github.com/PaulStoffregen/cores/blob/master/teensy4/AudioStream.h
+#include <arm_math.h>    // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h
```
Leaving a note here for later tests to confirm it works with the lib files in the repo instead.

ghostintranslation commented on filter_ladder.cpp:
```diff
@@ -0,0 +1,310 @@
+#include <Arduino.h>
+#include "filter_ladder.h"
+#include <math.h>
+#include <arm_math.h>
```
Same here.

ghostintranslation commented on filter_ladder.h:
```diff
@@ -0,0 +1,93 @@
+
+#include <Arduino.h>     // github.com/PaulStoffregen/cores/blob/master/teensy4/Arduino.h
+#include <AudioStream.h> // github.com/PaulStoffregen/cores/blob/master/teensy4/AudioStream.h
+#include <arm_math.h>    // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h
```
Same here.

ghostintranslation commented on math_helper.c:
This file does not exists in the original Audio lib

ghostintranslation commented on math_helper.h:
This file does not exists in the original Audio lib

ghostintranslation commented on sqrt_integer.c:
Let's keep the same structure as in the original lib, this file and the associated .h would go in a utility folder then.

ghostintranslation commented on synth_daisysp.h:
I can't find this file in the original lib?

ghostintranslation commented on synth_sine.h:
```diff
@@ -29,7 +29,8 @@
 #include "AudioStream.h" // github.com/PaulStoffregen/cores/blob/master/teensy4/AudioStream.h
-#include "arm_math/src/arm_math.h"    // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h
+//#include "arm_math/src/arm_math.h"    // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h
+#include <arm_math.h>    // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h
```
Leaving a note to test the local files again.

ghostintranslation commented on synth_tonesweep.cpp:
This file also has some changes in the original lib.

ghostintranslation commented on synth_waveform.cpp:
```diff
@@ -26,7 +26,8 @@
 #include "synth_waveform.h"
-#include "arm_math/src/arm_math.h" 
+//#include "arm_math/src/arm_math.h" 
+#include <arm_math.h> 
```
Leaving a note to test the local files instead.

ghostintranslation commented on synth_waveform.h:
```diff
@@ -29,7 +29,8 @@
 #include <AudioStream.h> // github.com/PaulStoffregen/cores/blob/master/teensy4/AudioStream.h
-#include "arm_math/src/arm_math.h"   // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h
+//#include "arm_math/src/arm_math.h"   // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h
+#include <arm_math.h>   // github.com/PaulStoffregen/cores/blob/master/teensy4/arm_math.h
```
same here

ghostintranslation commented on synth_wavetable.cpp:
```diff
@@ -0,0 +1,439 @@
+#include <Arduino.h>
+#include "synth_wavetable.h"
+#include <dspinst.h>
+//#include <SerialFlash.h>
```
Was this not needed? It is uncommented in the original lib.

ghostintranslation commented:
I finished reviewing the code, I will now proceed to testing.

rheslip commented:
Thanks for all the work fixing up my code. I'm not very good at coding and toolchains but hopefully I saved you some time porting all those modules.

Synth_daisySP etc was not part of the original Teensy Audio library. I added it so I could use the stuff in ElectroSmith's DaisySP DSP library within the Teensy Audio framework.


________________________________
From: Ghost In Translation ***@***.***>
Sent: July 20, 2025 9:08 PM
To: ghostintranslation/pico-audio ***@***.***>
Cc: rheslip ***@***.***>; Author ***@***.***>
Subject: Re: [ghostintranslation/pico-audio] Added missing functions (PR #6)

[https://avatars.githubusercontent.com/u/50341468?s=20&v=4]ghostintranslation left a comment (ghostintranslation/pico-audio#6)<https://github.com/ghostintranslation/pico-audio/pull/6#issuecomment-3094979532>

I finished reviewing the code, I will now proceed to testing.

—
Reply to this email directly, view it on GitHub<https://github.com/ghostintranslation/pico-audio/pull/6#issuecomment-3094979532>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABRTUUCWDRQCP54KEOXUSXD3JQ4PZAVCNFSM6AAAAAB56YM6PCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAOJUHE3TSNJTGI>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

ghostintranslation committed:
Fixed imports of local files for the arm_math lib and AudioStream + put back the macros for interrupts to limit changes compared to original lib + updated files compared original lib

ghostintranslation commented:
I pushed changes related to what we mentioned here, the PolySynth example works, but in the SimpleDrum I get an error of Serial being undefined and in the LadderFilter there is a problem with the arm_math lib.

@h4yn0nnym0u5e it looks like we do need the source files and not just the headers? But I'm not sure how to include them, just copying them won't make it compile them.

ghostintranslation committed:
Fixed compile linking issues with local copy of arm_math lib

ghostintranslation commented:
With the help of Gemini CLI I moved files and added the library.properties with some required flags to allow Arduino to take the arm_math files when compiling. 

I tested all examples, did minor edits, everything seem to work.

ghostintranslation committed:
Merge branch 'main' of github.com:ghostintranslation/pico-audio into rheslip/main

ghostintranslation commented:
@rheslip  @okyeron @h4yn0nnym0u5e @makingsoundmachines 
I'd appreciate if you could test and confirm everything works fine!

rheslip commented:
I tested some of my apps and tweeked a few of the Teensy examples to work on my hardware. Everything I tested works except for analyse_fft256 - its an easy fix:

The Teensy architecture test at line 65 of analyse_fft256.cpp is false so it doesn't compile the code in. Just cut and paste the test for pico2 from analyse_fft1024.cpp.

Thanks for cleaning this all up - this is a really nice library to have for the Pico2!
________________________________
From: Ghost In Translation ***@***.***>
Sent: July 24, 2025 8:51 PM
To: ghostintranslation/pico-audio ***@***.***>
Cc: rheslip ***@***.***>; Mention ***@***.***>
Subject: Re: [ghostintranslation/pico-audio] Added missing functions (PR #6)

[https://avatars.githubusercontent.com/u/50341468?s=20&v=4]ghostintranslation left a comment (ghostintranslation/pico-audio#6)<https://github.com/ghostintranslation/pico-audio/pull/6#issuecomment-3115438549>

@rheslip<https://github.com/rheslip> @okyeron<https://github.com/okyeron> @h4yn0nnym0u5e<https://github.com/h4yn0nnym0u5e> @makingsoundmachines<https://github.com/makingsoundmachines>
I'd appreciate if you could test and confirm everything works fine!

—
Reply to this email directly, view it on GitHub<https://github.com/ghostintranslation/pico-audio/pull/6#issuecomment-3115438549>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABRTUUCKHVRTWZUQIJJTUJT3KF5S5AVCNFSM6AAAAAB56YM6PCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCMJVGQZTQNJUHE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

rheslip committed:
Fixed architecture test in analyze_fft256, uodated readme

Fixed architecture test in analyze_fft256. Updated readme.md to reflect current state of library after cleanup by Ghostintranslation

ghostintranslation commented:
Change looks good.

I will wait to see if anyone else can test before merging.

ghostintranslation commented:
Merged to move on, if anything needs to be changed/fixed we can do more PRs.
